src/ComplexSelector.js

```diff
@@ -0,0 +1,128 @@
+import reduce from 'lodash/reduce';
+import split from 'lodash/split';
+
+export default class ComplexSelector {
+  constructor(buildPredicate, findWhereUnwrapped, childrenOfNode) {
+    this.buildPredicate = buildPredicate;
+    this.findWhereUnwrapped = findWhereUnwrapped;
+    this.childrenOfNode = childrenOfNode;
+  }
+
+  getSelectors(selector) {
+    const cleaned = selector.replace(/\s{2,}/g, ' ');
+    const selectors = split(cleaned, ' ');
+    return reduce(selectors, (memo, sel) => {
+      if (sel === '+' || sel === '~') {
+        const temp = memo.pop();
+        memo.push(sel, temp);
+        return memo;
+      }
+
+      memo.push(sel);
+      return memo;
+    }, []);
+  }
+
+  handleSelectors(selectors, wrapper) {
+    const recurseSelector = (offset, fn, pre) => {
+      const predicate = pre || this.buildPredicate(selectors[offset]);
+      const nextWrapper = this.findWhereUnwrapped(wrapper, predicate, fn);
+      const nextSelectors = selectors.slice(offset + 1);
+      return this.handleSelectors(nextSelectors, nextWrapper);
+    };
+
+    const buildSiblingPredicate = (first, second) => {
+      const firstPredicate = this.buildPredicate(first);
+      const secondPredicate = this.buildPredicate(second);
+
+      return (child) => {
+        if (firstPredicate(child)) {
+          return (sibling) => secondPredicate(sibling);
+        }
+
+        return false;
+      };
+    };
+
+    let { predicate, selectSiblings } = {};
+
+    if (selectors.length) {
+      switch (selectors[0]) {
+        case '>':
+          return recurseSelector(1, this.treeFilterDirect());
+        case '+':
+          predicate = buildSiblingPredicate(selectors[1], selectors[2]);
+          selectSiblings = (children, pre, results, idx) => {
+            const adjacent = children[idx + 1];
+            if (pre(adjacent)) { results.push(adjacent); }
+          };
+
+          return recurseSelector(2, this.treeFindSiblings(selectSiblings), predicate);
+        case '~':
+          predicate = buildSiblingPredicate(selectors[1], selectors[2]);
+          selectSiblings = (children, pre, results, idx) =>
+            children.slice(idx + 1).map(child =>
+              pre(child) ? results.push(child) : null
+            );
+
+          return recurseSelector(2, this.treeFindSiblings(selectSiblings), predicate);
+        default:
+          return recurseSelector(0);
+      }
+    }
+
+    return wrapper;
+  }
+
+  find(selector, wrapper) {
+    if (typeof selector === 'string') {
+      const selectors = this.getSelectors(selector);
+
+      return this.handleSelectors(selectors, wrapper);
+    }
+
+    const predicate = this.buildPredicate(selector);
+    return this.findWhereUnwrapped(wrapper, predicate);
+  }
+
+  treeFilterDirect() {
+    return (tree, fn) => {
+      const results = [];
+      this.childrenOfNode(tree).forEach(child => {
```

Comment:
Before in each of the Wrapper classes the `findWhereUnwrapped` always used `treeFilter`, which worked for looping through all nodes, but not for other use cases. That's why I need `treeFilterDirect` and `treeFindSiblings`